### DISCUSSION TITLE
Newbie trying to swap pdf viewer

### DISCUSSION THREAD

**Discussion by javadvilldu** [stephanrauh/ngx-extended-pdf-viewer#2519] [Q&A]
Completely new to web-development, I was just asked to swap our PDF-viewer to yours, mainly to get the possibility to use pinch-zoom on tablets/phones (I'm currently testing it on a Samsung tablet). Well, lets give it a try I thought. Now I'm stuck at multiple fronts.

I have managed to swap it, and the minimum use kinda works. But it crashes when zooming to aggressively in smaller size PDFs, which made me lock the `zoomMax `to 200%. But when stepping up the file sizes a bit (the last one I tried was 14.4MB), that's not enough, it still crashes. [Here](https://github.com/stephanrauh/ngx-extended-pdf-viewer/issues/669) I read about testing `pdfDefaultOptions.useOnlyCssZoom=true;`, but that doesn't seem to be available any longer. Not `pdfDefaultOptions.renderer='svg'` or `pdfDefaultOptions.enableWebGL=true` either, what I can tell. Any other way to get the same results?

When opening it on an iOS device instead, no PDFs is shown at all. I haven't even gotten to looking at this yet. Is it some obvious reason, an extra step that needs to be done or so, or does it need it's own digging?

Then some other issues with buttons not showing, but that is irrelevant if I can't fix the zooming/displaying issues..

Any help is appreciated.. :)

**Comment by stephanrauh** [comment]
I suggest you start experimenting with a fresh project first. I know that doesn't solve your task at hand, but it helps you to figure out which issues are caused by your application and which issues are my fault. Basically, that's only a matter of running `ng new` and `ng add ngx-extended-pdf-viewer`, so your sandbox application should be up and running within five minutes (see https://pdfviewer.net/extended-pdf-viewer/getting-started).

I've been trying to reproduce your zooming issue on my iPad. It works flawlessly, at least in my showcase. Most PDF files are smaller than 14 MB, so it might have to do with your particular PDF file. Is it a high-resolution file? I've got one file that renders extremely slow on resolutions above 200%: https://pdfviewer.net/extended-pdf-viewer/contextmenu. Maybe your PDF file simply has to many details.

It's also a good idea to watch the console window. Are the any error messages?

Another thing you should check: are you using the worker thread? In Google Chrome, you can check this in the "sources" tab by opening the "threads" panel and looking for "pdf.worker.*":
![image](https://github.com/user-attachments/assets/60ef3987-fdce-487c-b8c5-97187286eff9)

If you're not using the worker thread, the PDF viewer still works in a fall-back mode, but there's a heavy performance penalty.

**Reply by javadvilldu** [reply]
I have now tried it in a complete blank project and I still got the crashes when zooming to deep or fast. It got much better when backing to 20.0.2 and I could use `pdfDefaultOptions.useOnlyCssZoom=true;` and `pdfDefaultOptions.renderer='svg'`. I then locked it to 250% as max,

The problem running it on iOS disappeared when upgrading from iOS 16 to 17, or setting `[forceUsingLegacyES5]="true"`. Anything you recognize? By reading [this](https://pdfviewer.net/extended-pdf-viewer/browser-support), I get the feeling that I don't want it enabled constantly. Any good way of detecting situations when it may be neccesary, and in that case enable it? Should _(pdfLoadingFailed)_ be triggered when not being able to load the pdf in this scenario? Then perhaps I could set `[forceUsingLegacyES5]="true"` if that happens?

**Reply by stephanrauh** [reply]
I suspect you'll be able to reproduce your zooming issue on https://mozilla.github.io/pdf.js/web/viewer.html. That's the showcase of the PDF viewer of Firefox, which is the viewer I sort of hijacked. I've added quite a few minor modifications to the viewer, but I know next to nothing about the rendering engine, so I'm pretty sure I wasn't the one who messed it up. It's possible but unlikely.

Can you check this, please, and if it turns out to be an issue of the base library, would you mind to report the bug at https://github.com/mozilla/pdf.js/issues?

As for iOS: what about the console log messages? It's possible that iOS 16 isn't compatible any more, but if everything works as intended, ngx-extended-pdf-viewer detects automatically that you're not using a top-notch browser and activates `[forceUsingLegacyES5]`. Recently, I suspect the browser detection doesn't always work, and that the Babel configuration is broken and generates code incompatible to older iOS devices. You can see the effect of the browser detection in the network tab, and if the code is incompatible to your iOS version, it should show in the console.

And, yes, you're right: the ES5 code is larger and slower than the shiny modern code. That's why you don't want to use it unless you really need it.
